src/dataflow/src/sink/kafka.rs

```diff
@@ -237,283 +219,243 @@ where
+                if !should_emit {
+                    // Skip stale data for already published timestamps
+                    continue;
+                }
 
-                        if !should_emit {
-                            continue;
-                        }
+                if diff == 0 {
+                    // Explicitly refuse to send no-op records
+                    continue;
+                };
 
-                        queue.push_back((row, time, diff));
-                        if let Some(consistency) = &mut consistency {
-                            // Note that since a single differential message
-                            // turns into |diff| messages we need to increment
-                            // message counts by |diff| instead of just 1 for
-                            // the consistency topic
-                            let insert =
-                                consistency.update_timestamp_count(time, diff.abs() as i64);
-
-                            if insert {
-                                // Send a BEGIN message for a timestamp the first
-                                // time we encounter it
-                                consistency.queue.push_back((
-                                    SinkConsistencyState::Begin,
-                                    time,
-                                    None,
-                                ));
-                            }
-                        }
+                let diff_pair = if diff < 0 {
+                    DiffPair {
+                        before: Some(&row),
+                        after: None,
                     }
+                } else {
+                    DiffPair {
+                        before: None,
+                        after: Some(&row),
+                    }
+                };
+
+                let transaction_id = match connector.consistency {
+                    Some(_) => Some(time.to_string()),
+                    None => None,
+                };
+
+                let (key, value) = s.encoder.encode_unchecked(
+                    connector.key_schema_id,
+                    connector.value_schema_id,
+                    diff_pair,
+                    transaction_id,
+                );
+
+                // For diffs other than +/- 1, we send repeated copies of the
+                // Avro record [diff] times. Since the format and envelope
+                // capture the "polarity" of the update, we need to remember
+                // how many times to send the data.
+                let rows = pending_rows.entry(time).or_default();
+                rows.push(EncodedRow {
+                    key,
+                    value,
+                    count: diff.abs() as usize,
                 });
+                s.metrics.rows_queued.inc();
+            }
+        });
+
+        // Move any newly closed timestamps from pending to ready
+        let mut closed_ts: Vec<u64> = pending_rows
```

Comment:
I feel like BTreeMap's `drain_filter` does exactly what you want here but its nightly only :(

Reply:
It does; I look forward to replacing this whole bit eventually :/

Reply:
We could just copy the `drain_filter` code from std into our `ore`, it seems like less than a few dozen lines of code. Which is more than you've got here, but might allow for reuse/weighing in on the stabilization issue? Not sure if we would want it anywhere else, though.

Reply:
It relies a little too much on the internals of btreemap than I'm comfortable with - I don't really want to deal with keeping it in sync with upstream even if all the APIs it uses are public.

Reply:
yeah I agree with Eli here.